### PR TITLE
Refactor training script + CLI

### PR DESCRIPTION
**Pull request by smguo**
This PR merges the two training modules in the code base (vq_vae.py & vq_vae_supp.py merged to vae.py and run_training.py) and added a CLI to the training script. I try to the PR as backward compatible as possible and keep @miaecle 's latest changes to training script from `quick_fix`. Below are the changes that would affect training using the old script/parameters: 

1. Extend matching loss to contrastive loss (with negative weights for patches that do not belong to the same trajectory)
2. Relation matrix now stores the labels instead of weights. Weights are assigned to corresponding labels in the forward pass of the model
3. Matching loss is now computed as the mean of the weight matrix instead of the sum. This removes batch size dependence of matching loss (matrix sum would increase with increasing batch size). 

There are now 2 train functions in `run_training.py`: `train` for training VAE models without using data loaders for backward compatibilities and `train_with_loader` for training new models using data loaders. We can discuss if we want to rewrite the VAE models to use data loaders.

### CODE REVIEW

File: SingleCellPatch/generate_trajectories.py [hunk 1]
```diff
@@ -476,18 +480,17 @@ def patch_name_to_tuple(f):
                 assert (site, t_idx, trajectory[t_idx]) in patch_id_mapping, "Cannot find /%s/%d_%d" % (site, t_idx, trajectory[t_idx])
                 ref_patch_id = patch_id_mapping[(site, t_idx, trajectory[t_idx])]
                 patch_ids.append(ref_patch_id)
-                # Adjacent frames
-                if t_idx - 1 in t_ids:
-                    adj_patch_id = patch_id_mapping[(site, t_idx - 1, trajectory[t_idx - 1])]
-                    relations[(ref_patch_id, adj_patch_id)] = 2
+                labels[ref_patch_id] = label_count
```
Comment (bryantChhun) [inline SingleCellPatch/generate_trajectories.py:500]: It looks like the `labels[ref_patch_id] = label_count` assigns based on an increment per iteration, while `ref_patch_id` is assigned based on enumeration in `{patch_name_to_tuple(f): i for i, f in enumerate(fs)}`.  Is there a potential for inconsistent assignment if the order of `fs` is inconsistent?   Should we sort `fs` before assigning `patch_id_mapping`?

File: SingleCellPatch/generate_trajectories.py [hunk 2]
```diff
@@ -476,18 +480,17 @@ def patch_name_to_tuple(f):
...
                 if t_idx + 1 in t_ids:
                     adj_patch_id = patch_id_mapping[(site, t_idx + 1, trajectory[t_idx + 1])]
                     relations[(ref_patch_id, adj_patch_id)] = 2
-
+                    relations[(adj_patch_id, ref_patch_id)] = 2
```
Comment (bryantChhun) [inline SingleCellPatch/generate_trajectories.py:504]: If I understand correctly, the removed block:
```
if t_idx - 1 in t_ids:
        adj_patch_id = patch_id_mapping[(site, t_idx - 1, trajectory[t_idx - 1])]
        relations[(ref_patch_id, adj_patch_id)] = 2
```
is equivalent to the bottom line in this block:
```
if t_idx + 1 in t_ids:
        adj_patch_id = patch_id_mapping[(site, t_idx + 1, trajectory[t_idx + 1])]
        ...
        relations[(adj_patch_id, ref_patch_id)] = 2
```
which I believe if the boundary conditions (t_idx = 0, or t_idx = max) don't pose problems.

File: pipeline/segmentation.py [hunk 3]
```diff
@@ -92,9 +92,10 @@ def segmentation(raw_folder_: str,
 
 def instance_segmentation(raw_folder: str,
                           supp_folder: str,
-                          val_folder: str,
```
Comment (bryantChhun) [inline pipeline/segmentation.py:95]: for now we still need `val_folder` because `run_segmentation` passes the same number of arguments to each of its methods: `segmentation`, `instance_segmentation`, `segmentation_validation`

File: pipeline/patch_VAE.py [hunk 4]
```diff
@@ -113,13 +113,9 @@ def build_trajectories(summary_folder: str,
...
 def assemble_VAE(raw_folder: str,
                  supp_folder: str,
-                 # channels: list,
-                 # model_path: str,
                  sites: list,
                  config: YamlReader,
-                 # network: str = None,
-                 # save_mask: bool=False,
-                 # mask_channels: list=[-2, -1],
+                 patch_type: str='masked_mat',
```
Comment (bryantChhun) [inline pipeline/patch_VAE.py:119]: We'll have to be careful with the arguments passed to this method, because they are set by the worker in `run_vae`.  This can be adjusted in the next PR focused on `patch_vae.py`

File: pipeline/patch_VAE.py [hunk 5]
```diff
@@ -352,6 +343,7 @@ def process_VAE(raw_folder: str,
                 supp_folder: str,
                 sites: list,
                 config_: YamlReader,
+                gpu: int=0,
```
Comment (bryantChhun) [inline pipeline/patch_VAE.py:347]: Same here about args to worker processes.  `run_vae.py` sends only `(raw, supp, sites, config)` to its workers.

File: pipeline/patch_VAE.py [hunk 6]
```diff
@@ -392,13 +384,16 @@ def process_VAE(raw_folder: str,
...
 
     # these sites should be from a single condition (C5, C4, B-wells, etc..)
-    # model_path = os.path.join(weights_dir, 'model.pt')
-    # model_name = os.path.basename(weights_dir)
-    # output_dir = os.path.join(raw_folder, model_name)
-    # os.makedirs(output_dir, exist_ok=True)
-    # output_dir = config_.inference.weights
-    output_dir = raw_folder
+    model_dir = os.path.dirname(model_path)
+    model_name = os.path.basename(model_dir)
```
Comment (bryantChhun) [inline pipeline/patch_VAE.py:390]: We could add `model_name` to the config

File: run_training.py [hunk 7]
```diff
@@ -1342,106 +858,109 @@ def concat_relations(relations, offsets):
...
+        help='path to yaml configuration file'
+    )
+
+    return parser.parse_args()
+
+
+if __name__ == '__main__':
+    args = parse_args()
+    main(args.config)
+
```
Comment (bryantChhun) [inline run_training.py:967]: Overall I like `run_training.py`.  Later we should look into breaking out the cli parsing commands here from the torch-specific stuff like training and dataloading.

**Review by bryantChhun — APPROVED** [review]
Looks great!  Some questions for my own clarification.  We can iterate on smaller details like the training cli in later PRs.

File: SingleCellPatch/generate_trajectories.py (hunk 1 above)
Comment (smguo) [reply SingleCellPatch/generate_trajectories.py:500]: `fs` passed to this function has been sorted earlier in `asseble_VAE`, but the assignment should work even if `fs` is not sorted as long as `cell_id` is mapped correctly between `patches` and `trajectories`, which I'm not sure if it has been tested...

File: SingleCellPatch/generate_trajectories.py (hunk 2 above)
Comment (smguo) [reply SingleCellPatch/generate_trajectories.py:504]: Yes here the block loops through all available indices in `t_ids` and check if next time point is also in `t_ids`, so (t_idx = 0, or t_idx = max) wouldn't pose problems since they would simply get skipped if the next time point is not available. All the adjacent time points can be found by 1 block so the other one is actually redundant

File: pipeline/segmentation.py (hunk 3 above)
Comment (smguo) [reply pipeline/segmentation.py:95]: That's right. We should probably remove this constraints of the same number of arguments in future PRs. I found it's confusing that some arguments are passed to the function but not used.

File: pipeline/patch_VAE.py (hunk 4 above)
Comment (smguo) [reply pipeline/patch_VAE.py:119]: Yeah a quick solution is to make this parameter part of the config. I'll add a TODO note in the code.

File: pipeline/patch_VAE.py (hunk 5 above)
Comment (smguo) [reply pipeline/patch_VAE.py:347]: Good catch! I add gpu to the input argument of in workers.

File: pipeline/patch_VAE.py (hunk 6 above)
Comment (smguo) [reply pipeline/patch_VAE.py:390]: Sure. `model_name` here is only used to name the output folder so it can be user specified. We can make the default same as the name of the model folder (which is parsed from `model path`)

File: run_training.py (hunk 7 above)
Comment (smguo) [reply run_training.py:967]: I agree. `main` can use some more refactoring.
